src/repositories/versionable/VersionableRepository.ts

```diff
@@ -0,0 +1,161 @@
+import * as bcrypt from 'bcrypt';
+import * as mongoose from 'mongoose';
+export default class VersionableRepository
+< D extends mongoose.Document, M extends mongoose.Model<D>> {
+
+    private versionableModel: M;
+    constructor(model) {
+        this.versionableModel = model;
+    }
+
+    public async checkUnique(email: string) {
+        // to check that email field is unique
+        const result = await this.getAll({email}, 'email');
+        return ( result.length > 0 );
+    }
+
+    public async create(option, userid): Promise<D> {
+        const id = mongoose.Types.ObjectId();
+        const data = {
+            ...option,
+            _id: id,
+            createdAt: Date.now(),
+            createdBy: userid,
+            originalID: id,
+        };
+        const count = await this.checkUnique(data.email);
+        if (count) {
+            throw new Error('email exists');
+        }
+        else {
+            const result = await this.versionableModel.create(data);
+            return result.toObject({transform: (doc, ret) => {
+                delete ret.password;
+                return ret;
+            }});
+        }
+    }
+
+    public async createWithHash(options, userid): Promise<D> {
+        const saltCount = 10;
```

Comment:
Why are we doing this all logic inside repository?
We should be doing this inside controller.